File: JeMPI_Apps/JeMPI_Configuration/src/main/scala/configuration/CustomLibMPIExpandedGoldenRecord.scala
package configuration

import java.io.{File, PrintWriter}

private object CustomLibMPIExpandedGoldenRecord {

  private val classLocation = "../JeMPI_Shared_Source/custom"
  private val customClassName = "CustomLibMPIExpandedGoldenRecord"
  private val packageText = "org.jembi.jempi.libmpi.dgraph"

  def generate(fields: Array[Field]): Unit =
    val classFile: String = classLocation + File.separator + customClassName + ".java"
    println("Creating " + classFile)
    val file: File = new File(classFile)
    val writer: PrintWriter = new PrintWriter(file)
    val margin = 40
    writer.println(
      s"""package $packageText;
         |
         |import com.fasterxml.jackson.annotation.JsonInclude;
         |import com.fasterxml.jackson.annotation.JsonProperty;
         |import org.jembi.jempi.libmpi.MpiExpandedGoldenRecord;
         |import org.jembi.jempi.shared.models.CustomDemographicData;
         |import org.jembi.jempi.shared.models.GoldenRecord;
         |
         |import java.util.List;
         |
         |@JsonInclude(JsonInclude.Include.NON_NULL)
         |record $customClassName(
         |${" " * 6}@JsonProperty("uid") String uid,
         |${" " * 6}@JsonProperty("GoldenRecord.source_id") List<LibMPISourceId> sourceId,""".stripMargin)
    fields.foreach {
      case field =>
        val parameterName = Utils.snakeCaseToCamelCase(field.fieldName)
        val parameterType = (if field.isList.isDefined && field.isList.get then "List<" else "") +
          field.fieldType + (if field.isList.isDefined && field.isList.get then ">" else "")
        writer.println(s"""${" " * 6}@JsonProperty("GoldenRecord.${field.fieldName}") ${parameterType} $parameterName,""".stripMargin)
    }
    writer.println(
      s"""${" " * 6}@JsonProperty("GoldenRecord.patients") List<CustomLibMPIDGraphPatientRecord> patients) {
         |""".stripMargin)

    writer.print(
      """
        |   GoldenRecord toGoldenRecord() {
        |      return new GoldenRecord(this.uid(),
        |                              this.sourceId() != null
        |                                    ? this.sourceId().stream().map(LibMPISourceId::toSourceId).toList()
        |                                    : List.of(),
        |                              new CustomDemographicData(""".stripMargin)

    fields.zipWithIndex.foreach {
      (field, idx) =>
        writer.println(
          s"${" " * (if (idx == 0) 0 else 56)}this.${Utils.snakeCaseToCamelCase(field.fieldName)}()" +
            (if (idx + 1 < fields.length) "," else "));"))
    }
    writer.println("   }")

    writer.println(
      """
        |   MpiExpandedGoldenRecord toMpiExpandedGoldenRecord() {
        |      return new MpiExpandedGoldenRecord(this.toGoldenRecord(),
        |                                         this.patients().stream().map(CustomLibMPIDGraphPatientRecord::toMpiPatientRecord).toList());
        |   }
        |""".stripMargin)

    writer.println("}")
    writer.flush()
    writer.close()
  end generate

}

File: JeMPI_Apps/JeMPI_Shared_Source/custom/CustomLibMPIExpandedGoldenRecord.java
package org.jembi.jempi.libmpi.dgraph;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;

import org.jembi.jempi.shared.models.CustomGoldenRecord;
import org.jembi.jempi.libmpi.MpiExpandedGoldenRecord;

@JsonInclude(JsonInclude.Include.NON_NULL)
record CustomLibMPIExpandedGoldenRecord(@JsonProperty("uid") String uid,
                                        @JsonProperty("GoldenRecord.source_id") List<LibMPISourceId> sourceId,
                                        @JsonProperty("GoldenRecord.aux_id") String auxId,
                                        @JsonProperty("GoldenRecord.given_name") String givenName,
                                        @JsonProperty("GoldenRecord.family_name") String familyName,
                                        @JsonProperty("GoldenRecord.gender") String gender,
                                        @JsonProperty("GoldenRecord.dob") String dob,
                                        @JsonProperty("GoldenRecord.city") String city,
                                        @JsonProperty("GoldenRecord.phone_number") String phoneNumber,
                                        @JsonProperty("GoldenRecord.national_id") String nationalId,
                                        @JsonProperty("GoldenRecord.entity_list") List<CustomLibMPIDGraphEntity> dgraphEntityList) {


   CustomGoldenRecord toCustomGoldenRecord() {
      return new CustomGoldenRecord(this.uid(),
                                    this.sourceId() != null
                                       ? this.sourceId().stream().map(LibMPISourceId::toSourceId).toList()
                                       : List.of(),
                                    this.auxId(),
                                    this.givenName(),
                                    this.familyName(),
                                    this.gender(),
                                    this.dob(),
                                    this.city(),
                                    this.phoneNumber(),
                                    this.nationalId());
   }

   MpiExpandedGoldenRecord toMpiExpandedGoldenRecord() {
      return new MpiExpandedGoldenRecord(this.toCustomGoldenRecord(),
                                         this.dgraphEntityList()
                                             .stream()
                                             .map(CustomLibMPIDGraphEntity::toMpiEntity)
                                             .toList());
   }

}
